contracts/handlers/NativeTokenHandler.sol

```diff
@@ -65,21 +79,68 @@ contract NativeTokenHandler is IHandler, ERCHandlerHelpers {
+        optionalMessage             length uint256 bytes (96 + len(destinationRecipientAddress)) - (128 + len(destinationRecipientAddress))
+        optionalMessage                    bytes   bytes (160 + len(destinationRecipientAddress)) - END
      */
     function executeProposal(bytes32 resourceID, bytes calldata data) external override onlyBridge returns (bytes memory) {
-        (uint256 amount) = abi.decode(data, (uint256));
+        uint256 amount;
+        uint256 lenDestinationRecipientAddress;
+
+        (amount, lenDestinationRecipientAddress) = abi.decode(data, (uint256, uint256));
+        address recipientAddress = address(bytes20(bytes(data[64:64 + lenDestinationRecipientAddress])));
+
         address tokenAddress = _resourceIDToTokenContractAddress[resourceID];
-        address recipientAddress = address(bytes20(bytes(data[64:84])));
-        uint256 convertedAmount = convertToExternalBalance(tokenAddress, amount);
 
-        (bool success, ) = address(recipientAddress).call{value: convertedAmount}("");
-        if(!success) revert FailedFundsTransfer();
-        emit FundsTransferred(recipientAddress, amount);
+        // Optional message recipient transformation.
+        uint256 pointer = 64 + lenDestinationRecipientAddress;
+        uint256 gas;
+        uint256 messageLength;
+        OptionalMessageCheck optionalMessageCheck;
+        if (data.length > (pointer + 64)) {
+            (gas, messageLength) = abi.decode(data[pointer:], (uint256, uint256));
+            pointer += 64;
+            if (gas > 0 && messageLength > 0 && (messageLength + pointer) <= data.length) {
+                optionalMessageCheck = OptionalMessageCheck.Valid;
+                if (recipientAddress == transformRecipient) {
+                    recipientAddress = _defaultMessageReceiver;
+                }
+            } else {
+                gas = defaultGas;
+                optionalMessageCheck = OptionalMessageCheck.Invalid;
+            }
+        }
+
+        if (optionalMessageCheck == OptionalMessageCheck.Invalid) {
+            return abi.encode(
+                tokenAddress,
+                recipientAddress,
+                amount,
+                abi.encode(false, abi.encodeWithSignature("InvalidEncoding()"))
+            );
+        }
+
+        uint256 externalAmount = convertToExternalBalance(tokenAddress, amount);
+        bytes memory recipientMessage = "";
+        if (optionalMessageCheck == OptionalMessageCheck.Valid) {
+            bytes memory message = bytes(data[pointer:pointer + messageLength]);
+            recipientMessage = abi.encodeWithSelector(
+                ISygmaMessageReceiver(recipientAddress).handleSygmaMessage.selector,
+                tokenAddress,
+                externalAmount,
+                message
+            );
+        }
+        (bool success, bytes memory result) =
```

Comment:
There are some slight differences between implementation which would be solved by extracting it. 

Here I assume this check should be in the if-case.

Reply:
This particular check has to be here because we transfer the native token and need to make sure it is received successfully.